lib/ecto/migrator.ex

```diff
@@ -153,41 +144,65 @@ defmodule Ecto.Migrator do
   end
 
   defp do_down(repo, version, module, opts) do
-    run_maybe_in_transaction(repo, module, fn ->
+    async_migrate_maybe_in_transaction(repo, version, module, :down, opts, fn ->
       attempt(repo, version, module, :forward, :down, :down, opts)
         || attempt(repo, version, module, :backward, :change, :down, opts)
         || {:error, Ecto.MigrationError.exception(
             "#{inspect module} does not implement a `down/0` or `change/0` function")}
     end)
-    |> case do
-      :ok ->
-        verbose_schema_migration repo, "update schema migrations", fn ->
-          SchemaMigration.down(repo, version, opts[:prefix])
-        end
-        :ok
-      error ->
-        error
-    end
   end
 
-  defp run_maybe_in_transaction(repo, module, fun) do
-    fn -> do_run_maybe_in_transaction(repo, module, fun) end
-    |> Task.async()
-    |> Task.await(:infinity)
+  defp async_migrate_maybe_in_transaction(repo, version, module, direction, opts, fun) do
+    parent = self()
+    ref = make_ref()
+
+    %{pid: pid} = task =
+      Task.async(fn -> run_maybe_in_transaction(parent, ref, repo, module, fun) end)
+
+    try do
+      receive do
+        {^ref, :ok} ->
+          verbose_schema_migration repo, "update schema migrations", fn ->
+            apply(SchemaMigration, direction, [repo, version, opts[:prefix]])
+          end
+
+        {^ref, _} ->
+          :ok
+
+        {:EXIT, ^pid, _} ->
+          :ok
+      end
+    catch
+      kind, error ->
+        Task.shutdown(task, :brutal_kill)
+        :erlang.raise(kind, error, System.stacktrace())
+    else
+      _ ->
+        send(task.pid, ref)
+        Task.await(task, :infinity)
+    end
   end
 
-  defp do_run_maybe_in_transaction(repo, module, fun) do
-    cond do
-      module.__migration__[:disable_ddl_transaction] ->
-        fun.()
-      repo.__adapter__.supports_ddl_transaction? ->
-        {:ok, result} = repo.transaction(fun, log: false, timeout: :infinity)
-        result
-      true ->
-        fun.()
+  defp run_maybe_in_transaction(parent, ref, repo, module, fun) do
+    if module.__migration__[:disable_ddl_transaction] ||
+         not repo.__adapter__.supports_ddl_transaction? do
+      send_and_receive(parent, ref, fun.())
+    else
+      {:ok, result} =
+        repo.transaction(
+          fn -> send_and_receive(parent, ref, fun.()) end,
+          log: false, timeout: :infinity
+        )
+
+      result
     end
   catch kind, reason ->
-    {kind, reason, System.stacktrace}
+    send_and_receive(parent, ref, {kind, reason, System.stacktrace})
+  end
+
+  defp send_and_receive(parent, ref, value) do
```

Comment:
What do you think about naming this `report_and_wait_for_schema_migrations_update` or similar? It can also be splitted into: 

```elixir
report_migration_result(parent, {ref, value})
wait_for_schema_migrations_update(ref, value)
```

I'm trying to be more clear about why this "two-step" flow is required. With that goal in mind, the code above:

```elixir
try do
  receive do
    {^ref, :ok} ->
      verbose_schema_migration repo, "update schema migrations", fn ->
        apply(SchemaMigration, direction, [repo, version, opts[:prefix]])
      end
      
    {^ref, _} ->
      :ok
      
    {:EXIT, ^pid, _} ->
      :ok
  end
catch
  kind, error ->
    Task.shutdown(task, :brutal_kill)
    :erlang.raise(kind, error, System.stacktrace())
else
  _ ->
    send(task.pid, ref)
    Task.await(task, :infinity)
end
```

could be abstracted to a:

```elixir
parent = self()
result_ref = make_ref()

task = Task.async(fn -> run_maybe_in_transaction(parent, result_ref, repo, module, fun) end)

case wait_for_migration_result(task, result_ref) do
  :ok ->
    # The table with schema migrations can only be updated from the parent process
    # because it has a lock acquired on that table.
    report_schema_migrations_update(task, fn ->
      verbose_schema_migration repo, "update schema migrations", fn ->
        apply(SchemaMigration, direction, [repo, version, opts[:prefix]])
      end
    end)
  
  _ ->
    :ok
end)

Task.await(task, :infinity)
```

Thoughts?

Reply:
I like it. I pushed something similar, although I kept the try catch inline because I don't want to hide the task management into a bunch of functions and instead keep it all in one place. :)